api/buckets/service.go

```diff
@@ -128,42 +149,248 @@ func (s *Service) createBucket(ctx context.Context, dbID thread.ID, dbToken thre
+		return nil, err
 	}
-	buckPath, err := s.IPFSClient.Object().AddLink(ctx, path.IpfsPath(bootCid), bucks.SeedName, seedPth)
+	return path.IpfsPath(n.Cid()), nil
+}
+
+// newDirWithNode returns a new proto node directory wrapping the node,
+// which is encrypted if key is not nil.
+func newDirWithNode(n ipld.Node, name string, key []byte) (ipld.Node, error) {
+	dir := unixfs.EmptyDirNode()
+	dir.SetCidBuilder(dag.V1CidPrefix())
+	if err := dir.AddNodeLink(name, n); err != nil {
+		return nil, err
+	}
+	return encryptNode(dir, key)
+}
+
+// encryptNode returns the encrypted versoin of node if key is not nil.
+func encryptNode(n *dag.ProtoNode, key []byte) (*dag.ProtoNode, error) {
+	if key == nil {
+		return n, nil
+	}
+	cipher, err := encryptData(n.RawData(), key)
+	if err != nil {
+		return nil, err
+	}
+	en := dag.NodeWithData(unixfs.FilePBData(cipher, uint64(len(cipher))))
+	en.SetCidBuilder(dag.V1CidPrefix())
+	return en, nil
+}
+
+type treeNode struct {
+	name string
+	node ipld.Node
+}
+
+// newDirFromExistingPath returns a new dir based on path.
+// If key is not nil, this method recursively walks the path, encrypting files and directories.
+// If add is not nil, it will be included in the resulting (possibly encrypted) node under a link named addName.
+// This method returns the root node and a list of all new nodes (which also includes the root).
+func (s *Service) newDirFromExistingPath(ctx context.Context, pth path.Path, key []byte, add ipld.Node, addName string) (ipld.Node, []ipld.Node, error) {
+	rn, err := s.IPFSClient.ResolveNode(ctx, pth)
+	if err != nil {
+		return nil, nil, err
+	}
+	top, ok := rn.(*dag.ProtoNode)
+	if !ok {
+		return nil, nil, dag.ErrNotProtobuf
+	}
+	if key == nil {
+		nodes := []ipld.Node{top}
+		if add != nil {
+			if err := top.AddNodeLink(addName, add); err != nil {
+				return nil, nil, err
+			}
+			nodes = append(nodes, add)
+		}
+		return top, nodes, nil
+	}
+
+	// Walk the node, encrypting the leaves and directories
+	tree := make(map[cid.Cid]*treeNode)
+	var addNode *treeNode
+	if add != nil {
+		addNode = &treeNode{
+			name: addName,
+			node: add,
+		}
+	}
+	root, err := s.encryptTree(ctx, tree, top, "", key, addNode)
 	if err != nil {
-		return nil, fmt.Errorf("adding seed to prepared bucket: %s", err)
+		return nil, nil, err
 	}
-	if err = s.IPFSClient.Pin().Add(ctx, buckPath); err != nil {
-		return nil, fmt.Errorf("pinning prepared bucket: %s", err)
+
+	// Collect new nodes
+	nodes := make([]ipld.Node, len(tree))
+	i := 0
+	for _, tn := range tree {
+		nodes[i] = tn.node
+		i++
 	}
-	return buckPath, nil
+	return root.node, nodes, nil
+}
+
+// encryptTree walks all nodes in n, encrypting nodes and re-creating the dag structure.
+// add will be added to the encrypted root node if not nil.
+// tree will contain a map of all new nodes.
+func (s *Service) encryptTree(ctx context.Context, tree map[cid.Cid]*treeNode, root ipld.Node, name string, key []byte, add *treeNode) (*treeNode, error) {
```

Comment:
This is the meat of the method that encrypts an entire remote Unixfs dir while pulling it from the network